### PR TITLE
Expose the overlay from ApiExt

### PR DESCRIPTION
**Pull request by liuchengxu**
Expose the necessary API `overlay` from ApiExt for supporting the extrinsic execution proof.

### CODE REVIEW

**Review by cheme — COMMENTED** [review]
I just read #10783. IIUC, checking a extrinsic proof will require:
- block B hash, state root and extrinsic proof for for this block
- block B+1 hash and state root
- state root at extrinsic N (probably stored in state for every extrinsic), so need proof for it in block B+1, same for state root at N +1
- proof of changed value at N to be able to build and verify the initial overlay change: this is not optimal, ideally it will be only value that the extrinsic need to access, but with current code building root at N+1 will access all values (build from scratch, not iteratively).
Then run and check that the produced root at N+1 is right.

And here to build the proof of the initial change set (changed value at N), looks a bit tricky.
I think it would need extract the change set for 0 to N-1 extrinsic.
Then build a partially in memory trie state (triedbmut with a db that write in an inmemory layer and read in block b state).
Then create the proof for the initial change set by by querying all changes of 0 to N-1 on this temporary trie.

In the end, if I compare to just using a proof of extrinsic 0 to N, we would remove read access of unmodified value for 0 to N-1. Not sure how relevant it is for your runtime (maybe proof of extrinsic 0 to N is just easier).

I may have misunderstood some points, anyway, I don't think these will requires any other substrate changes (except execution verification).

**Comment by liuchengxu** [comment]
Thank you for taking the time to read the background! @cheme 

>checking a extrinsic proof will require:
>
>- block B hash, state root and extrinsic proof for for this block
>- block B+1 hash and state root
>- state root at extrinsic N (probably stored in state for every extrinsic), so need proof for it in block B+1, same for state root at N +1
>- proof of changed value at N to be able to build and verify the initial overlay change: this is not optimal, ideally it will be only >value that the extrinsic need to access, but with current code building root at N+1 will access all values (build from scratch, not iteratively).
> Then run and check that the produced root at N+1 is right.

Hmm, not sure if we're thinking about the same problem, I don't understand why block B+1 is involved, the execution proof of an extrinsic in block B is only related to the block B.

First of all, we need to know the verifier of the extrinsic execution proof does not own the whole state so he needs the proof to restore the essential state/trie_backend for executing the extrinsic. Checking the execution proof of an extrinsic then requires:
- the state_root before executing the extrinsic `pre_state_root`
- the state root after executing the extrinsic `post_state_root`
- the extrinsic itself
- the runtime code for executing the method `BlockBuilder_apply_extrinsic` with the call data(extrinsic)

Given an execution proof, we first restore a trie backend that contains all the storage values necessary for executing this extrinsic, apply the extrinsic by running the executor with the method `BlockBuilder_apply_extrinsic` and call data, and then calculate the post state root. If the locally produced `post_state_root` matches the one received along with the proof, we can say the proof is correct.

>probably stored in state for every extrinsic

Yes, that's what we're doing right now(we might optimize this later by not storing the intermediate storage root in the state). Now we forked the executive pallet to store the storage root after applying each extrinsic, https://github.com/paritytech/substrate/issues/10711 is related.

Based on my current understanding(may change with mode in-depth experiment), the required substrate changes could be:

- Expose the overlay (accumulated changeset for 0..N-1) before applying the extrinsic N so that I can generate the execution proof for extrinsic N https://github.com/subspace/subspace/blob/cc67dd39a3/cumulus/client/block-builder/src/lib.rs#L212-L219
- Conver the `OvelayedChanges` to a storage proof and merge it with the execution proof produced by `execution_proof`.https://github.com/subspace/subspace/blob/cc67dd39a3/cumulus/client/block-builder/src/lib.rs#L270-L271 (I'm still investigating).
- Return the `post_state_root` in `execution_proof_check` as wellhttps://github.com/paritytech/substrate/discussions/10987
- A method in `ProofProvider` to verify the proof produced by `executoin_proof`.(not 100% sure about this, we'll see)

**Comment by cheme** [comment]
> I don't understand why block B+1 is involved

To verify value for `pre_state_root` and `post_state_root` (if they are in state).

I did think a bit about it and, what would be the perfect situation would be a proof from state B-N to state B-N+1 with B being block number and N the extrinsic number.
Alas that would be a heck of a hacking, I don't think it is easily doable (the way we calculate storage root in state machine is always from B-0).
So we have a proof from B-0 with a verified execution state at B-N (state checked against `pre_state_root`) to B-N+1, then checked against `post_state_root`. Note that is is a bit implementation dependant (the way storage_root access state between two different calls may not be well specified enough).

**Comment by cheme** [comment]
What you could do is extract the trie node changes when applying state transition from block initialization and extrinsic 0 to N (transaction field of StorageChange after calling into_storage_changes).
Then you prove your execution over a trie backend that use as a storage the storage from B and an in memory overlay of changes from you extracted changes.

**Comment by liuchengxu** [comment]
>actually B0 is a bit misleading, it is really B (since at extrinsic 0 start there will already be some changes).

True, so we are actually storing the intermediate storage roots for the hooks and extrinsics:

- after the initialize_block
- after applying each extrinsic
- after the finalize_block (which is actually the current `state_root` field in the header)

Thanks for all the info! @cheme Overall I think we're on the same page, just some implementation details. In order to generate the execution proof for the extrinsic position at N+1 in a block B<sub>N+1</sub>, the core part is to prepare the state B<sub>N</sub>(the state after applying the extrinsic N), which is B<sub>0</sub> + Overlay<sub>0..N</sub>. As long as we have the state B<sub>N</sub>, we can precisely derive the storage proof for executing extrinsic N+1, so the deltas will be only from B<sub>N</sub>, not from B<sub>0</sub>.

The problem is that the proof genetrated this way will also [require the Overlay to be passed into the underlying `exeution_proof_check`](https://github.com/subspace/subspace/blob/cc67dd39a3/cumulus/client/block-builder/src/lib.rs#L271), which is a thing we'd like to get rid of since the verifier of this proof does not have to know the concept of overlay, he only needs a full storage proof to restore a trie backend to execute this extrinsic.

Therefore, we are thinking about a custom overlay which is also able to record the storage accessed during the extrinsic execution(replace it with the custom overlay [here](https://github.com/subspace/subspace/blob/cc67dd39a3/cumulus/client/block-builder/src/lib.rs#L218)). At the end of the day, the verify can check the execution proof with only a full complete storage proof including the necessary overlay changeset. Any suggestion for building a custom overlay which is able to record the storage just like what the proof recorder does?

**Comment by cheme** [comment]
Don't get me wrong, I got no issue with exposing the overlay as this PR propose.

Just pointing out the design is difficult to put in place.

> after applying each extrinsic

Storing the root after extrinsic, is the same as calculating at the end of the extrinsic: it needs to be in the execution to be able to check BN+1, so it will still include all changes from B0 to BN.

> The problem is that the proof genetrated this way will also require the Overlay to be passed into the underlying exeution_proof_check, which is a thing we'd like to get rid of since the verifier of this proof does not have to know the concept of overlay, he only needs a full storage proof to restore a trie backend to execute this extrinsic.

In this PR it seems right to use the overlay, as it is the actual state of execution and in a more general context it makes sense to chain multiple executions.
In your api, it should be a key option<value> map (and n child trie key option<value> map) that is basically the changeset at BN.
I think there is a struct for it in state_machine (StorageChange maybe).
This changeset obviously need to be verified.

Using these two proofs, will verify extrinsic N. Just it is a double proof: a proof of state machine state at BN (overlay) and a proof of execution at B0.
In the end it can be a lot smaller than a proof from B0 to BN because the first proof uses a registered root. So it removes all read access from B0 to BN.

The proof for the change set is a difficult story (I describe it in my first comment) and may prove pretty redundant with the proof of execution (could probably be optimize, but all this layers of things do not feel really safe).

So the design with two proofs only make sense if in the extrinsic 0 to N-1 you got lot of read accesses and few write accesses.

Generally my last comment suggest an approach that looks safer to produce a single proof (no need for a changeset).

In my last comment I suggest that the execution proof creation uses a TrieBackendStorage that is the trie encoded nodes from state B (usually rocksdb) completed from the trie encoded nodes from execution of block of size N-1 without finalization. This way you got a proof that register node from state BN and no content needed in overlay.


Of course when I say 2 or 1 proof, I omit the proof to check the roots of BN and BN+1.

> Therefore, we are thinking about a custom overlay which is also able to record the storage accessed during the extrinsic execution

Maybe it is related to the way you plan to merge the two proofs, or do some deltas of accesses between 0 to N and N.

I actually updated some old code doing it in some experimental branch yesterday (trying to lookup for access conflict between extrinsic), also previous change trie did it for write access.
But I would not be positive about such code getting merge.

>  the core part is to prepare the state BN(the state after applying the extrinsic N), which is B0 + Overlay0..N. As long as we have the state BN, we can precisely derive the storage proof for executing extrinsic N+1, so the deltas will be only from BN, not from B0

What I did suggest: using registered node (execution proof) at BN, apply them on the trie storage (by using a in memory overlay and implement trait TrieBackendStorage), is exactly 'prepare the state BN'. Then create execution proof on this prepared state to get a single proof without overlay.

Deriving a BN proof from B0 + Overay0..N seems quite tricky.

**Comment by cheme** [comment]
Another advantage of extranction execution proof by running on a true state BN (build from state B and overlay of changed trie node from 0 to BN) is that the call to storage root will only run on delta from BN to BN+1 (not 0 to BN+1).

**Comment by liuchengxu** [comment]
Yes, creating a true state BN also looks more straightforward to me! I can be dumb, could you please elaborate on how to convert the B0+overlay to a trie backend for the state BN? Convert the state of block B to a trie backend and conver the overlay a trie backend, and then merge the two into one trie backend for the state BN? @cheme 

Additionally, I believe we'll still need to add an optional argument `Option<TrieBackend>` so that we could pass in the state BN.

**Comment by cheme** [comment]
It's not trivial, I don't have all the code in mind, but basically in https://github.com/paritytech/substrate/blob/8df8d908c4d77a8dd19751784b6aca62159ddda8/primitives/state-machine/src/lib.rs#L594 we want to use a TrieBackend for BN.
That could be build from the trie backend at B by replacing in https://github.com/paritytech/substrate/blob/8df8d908c4d77a8dd19751784b6aca62159ddda8/primitives/state-machine/src/trie_backend_essence.rs#L76 field root by root at BN and replacing storage by one with the diff.
something like:
```
struct NodeStorageWithDelta<S> {
  inner: S,
  delta: HashMap<Hash, Option<EncodedNode>>,
}
impl TrieBackendStorage for NodeStorageWithDelta {
  fn get {
// check delta first, if undefined call inner
}
}
```
with Hash = Vec<u8> (it is actually prefixed) and EncodedNode = Vec<u8>
Where delta contains the encoded node change from the processing of initialize_block and all extrinsic before N.
Actually delta does not need to contain removed node.
And inner is S at B (from the trie backend at B).

Then the question is how to get the root at BN and the encoded node delta at BN, here I would just call `into_storage_changes` on the overlay after execution, the fields to use  on the result are:
```
	pub transaction: Transaction,
	/// The storage root after applying the transaction.
	pub transaction_storage_root: H::Out,
```

**Comment by liuchengxu** [comment]
@cheme Can be great if you could take a quick look if this is on the right track https://github.com/subspace/substrate/compare/8ce691dcc163c4ca3fe90a33f87c8821d1a73412...subspace:3f76ce1ac913855ba0c427a32b1807bec80fa310, it compiles but runs with panic, something is definitely wrong, I'm still on the debugging :D

**Comment by cheme** [comment]
another small thing I notice, in : 
```
			let current_state_backend = sp_state_machine::CurrentStateBackend::new(
				trie_backend,
				storage_changes.transaction,
				post_delta_root,
			);
```
it should be

```
			let current_state_backend = sp_state_machine::CurrentStateBackend::new(
				trie_backend,
				storage_changes.transaction,
				storage_changes.trasaction_storage_root,
			);
```
Actually for delta, maybe just passing `StorageChanges` struct could make sense (can be seen as previous transition result).

**Comment by liuchengxu** [comment]
One thing I noticed is `parent_hash` in `into_storage_changes` is unused in fact, https://github.com/paritytech/substrate/blob/6bbc6885509430aac4b0cba4ac6c1a11fe525307/primitives/state-machine/src/overlayed_changes/mod.rs#L503-L517  can we remove this argument in `into_storage_changes` then?

**Comment by liuchengxu** [comment]
If I understand correctly, instead of using a dummy backend, I should just use the state at B https://github.com/subspace/substrate/blob/7da1cf8348/client/service/src/client/call_executor.rs#L318

```
error[E0308]: mismatched types
   --> client/service/src/client/call_executor.rs:323:5
    |
323 |                 storage_changes.transaction,
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `memory_db::MemoryDB`, found associated type
    |
    = note:       expected struct `memory_db::MemoryDB<<<Block as BlockT>::Header as HeaderT>::Hashing, memory_db::HashKey<<<Block as BlockT>::Header as HeaderT>::Hashing>, Vec<u8>, memory_db::malloc_size_of::NoopTracker<Vec<u8>>>`
            found associated type `<<B as sc_client_api::Backend<Block>>::State as sc_client_api::StateBackend<<<Block as BlockT>::Header as HeaderT>::Hashing>>::Transaction`
    = help: consider constraining the associated type `<<B as sc_client_api::Backend<Block>>::State as sc_client_api::StateBackend<<<Block as BlockT>::Header as HeaderT>::Hashing>>::Transaction` to `memory_db::MemoryDB<<<Block as BlockT>::Header as HeaderT>::Hashing, memory_db::HashKey<<<Block as BlockT>::Header as HeaderT>::Hashing>, Vec<u8>, memory_db::malloc_size_of::NoopTracker<Vec<u8>>>`
    = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html

error[E0308]: mismatched types
   --> client/service/src/client/call_executor.rs:323:5
    |
323 |                 storage_changes.transaction,
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `memory_db::MemoryDB`, found associated type
    |
    = note:       expected struct `memory_db::MemoryDB<<<Block as BlockT>::Header as HeaderT>::Hashing, memory_db::HashKey<<<Block as BlockT>::Header as HeaderT>::Hashing>, Vec<u8>>`
            found associated type `<<B as sc_client_api::Backend<Block>>::State as sc_client_api::StateBackend<<<Block as BlockT>::Header as HeaderT>::Hashing>>::Transaction`
    = help: consider constraining the associated type `<<B as sc_client_api::Backend<Block>>::State as sc_client_api::StateBackend<<<Block as BlockT>::Header as HeaderT>::Hashing>>::Transaction` to `memory_db::MemoryDB<<<Block as BlockT>::Header as HeaderT>::Hashing, memory_db::HashKey<<<Block as BlockT>::Header as HeaderT>::Hashing>, Vec<u8>>`
```

Then should I convert `StateBackend::Transaction` to something that implements `HashDB` or a HashMap of trie nodes so that I can implement `TrieBackendStorage`? Use [this branch](https://github.com/subspace/substrate/tree/fraud-proof) if you'd like to compile it. @cheme

**Comment by cheme** [comment]
> If I understand correctly, instead of using a dummy backend, I should just use the state at B

yes, you can, even if I would make the call to `into_storage_change` out of prove_execution, and pass an optional memorydb and a root as param, but that's a detail.
Not just a detail as transaction to memorydb is certainly easier out of this function where you will probably have access to the type of Transaction (MemoryDb) instead of just its trait.

Other type than MemoryDb could be just a HashMap<H::Out, Vec<u8>> (prefix can be ignored).

**Comment by liuchengxu** [comment]
>Not just a detail as converting transaction to memorydb is certainly easier out of this function where you will probably have access to the type of Transaction (MemoryDb) instead of just its trait.

:+1: This way works.

@cheme Please have a look at the latest changes where I changed the input to `PrefixedMemoryDB` https://github.com/subspace/substrate/compare/8ce691dcc163c4ca3fe90a33f87c8821d1a73412...subspace:142eeeb5545a03edd69d99cae0d39fc0e8a64973 I have tested the execution proof locally, seemingly works well.

If the changes look good to you, I'd love to split this PR into multiple ones to make some progress incrementally:

1. Add an extra argument `delta_changes` to `execution_proof` for generating an execution proof for any extrinsic.
2. Support `verify_execution_proof` in `ProofProvider`.
3. Make `execution_proof` more flexible, e.g., the runtime code will be fetched another way in Subspace.
4. Check if any features missing for generating the execution proof for any stages(initialize_block, extrinsic, fianlize_block).

**Comment by cheme** [comment]
Sure no issue with that. Probably will try to get as less specific code as possible in substrate, but an additional parameter and the verify function sounds good to me.

**Comment by liuchengxu** [comment]
@cheme I found that the intermediate storage roots for the extrinsics collected after [`initialize_block`](https://github.com/subspace/subspace/blob/a86583508b/cumulus/pallets/executive/src/lib.rs#L242) and [`apply_extrinsic`](https://github.com/subspace/subspace/blob/a86583508b/cumulus/pallets/executive/src/lib.rs#L359) don't accordingly match the `post_delta_root`. I construct the delta changes using a custom BlockBuilder when generating an extrinsic execution proof, the re-built header is the same with the originally built one. I also inspected the OverlayedChanges in the originally built and later re-built process, which are the same as well, but the storage roots does not match. Any clue why this happens? (should be unrelated to #11019 whose main goal is to create a true state for some extrinsic and then verify the proof without overlay, I have verified this process works as expected).

**Comment by cheme** [comment]
Not sure if it is your issue, but calculated storage root are stored at each call (using a trait storage means it will be in next storage root calculation) so they are part of next extrinsic root.
Basically for extrinsic N-1:
- extrinsic N is called on state N-1', produce a root N that contains all root from N to N-1.
- root is stored making state N'
When building proof same happen, and into_storage_change will execute storage proof again (insertion or root N invalidate cache and the root is calculated again), returning root at N'.
When you check you therefore run on N' and things would work well.
But the root that are stored in state will not be the N' roots but the N roots.
I don't really think immediately of a good solution, ideally root will not be stored between extrinsic but only in finalize so N == N', but I don't know of a way to do it (I remember transient storage between extrinsic always been a bit of a discussion, but not sure if new things where added).
Maybe calling storage root at start of extrinsic instead of after (and at start of finalize) will be better:
- state N, storage root is called storing N
- rest of extrinsic is executed.
Yes can be the solution here.

**Comment by liuchengxu** [comment]
>Maybe calling storage root at start of extrinsic instead of after (and at start of finalize) will be better

@cheme This saved my day! Now the collected intermediate storage roots for extrinsics match the post_delta_root manually constructed later.

**Comment by liuchengxu** [comment]
Okay, previously I tested the whole process under the test environment where I can access the concrete type of `B::Transaction`, however, when it comes to the proper implementation, I still face the problem that how I can convert `B::Transaction` to a concrete struct MemoryDB. Any thoughts? @cheme

**Comment by cheme** [comment]
There is definitely a time when transaction is not a opaque associated type.
When running import it is in client/db code when writing the delta, in test I am not sure. In your case it is possible you still need to call into_storage_proof a bit higher on the calling stack. It is also possible that with the code you use the transaction/overlay payload is dropped before and some changes are required.
Also the concrete struct may not be a PrefixedMemoryDb but a MemoryDb, using memory db in the overlay will work (just ignore prefix parameter for get). May be another type, but should be doable to rebuild the db from it.

**Comment by liuchengxu** [comment]
Any insights for https://github.com/paritytech/substrate/discussions/10987 which is related to the second step I planned in https://github.com/paritytech/substrate/pull/10922#issuecomment-1065251109 ? @cheme

**Comment by cheme** [comment]
Your proof is state at BN so you execute on your proof extrinsic execution:
this only show that extrinsic can be applied on state BN.
So you also need to compare to BN+1 root: meaning you need to really run/prove extrinsic execution, and `note_finished_extrinsic` and a call to storage_root (the call matches the one done at start of next extrinsic).
Then you can just check that the overlay contains BN+1 root (rosult of `storage_root`)..
Or you can skip calling storage_root and call `into_storage_root` on overlay (which itself internally call storage_root).

Alternatively (may be cleaner): do an extrinsic that wraps call_extrinsic, note_finished_extrinsic and return result to call to storage_root.
Yes last solution is really cleaner, but requires new calls, which may not be very practical (actually only one if you method name as first param).

**Comment by cheme** [comment]
> @cheme Do you mean something like this [subspace/subspace@951e603#diff-7ef7237bf81649e2291cfa38414ce9519b21cf5cfb9e626fb31a5cdc9bcf6367R362-R374](https://github.com/subspace/subspace/commit/951e6038b71a42653a1e0eb14932af1fbcebaf3c#diff-7ef7237bf81649e2291cfa38414ce9519b21cf5cfb9e626fb31a5cdc9bcf6367R362-R374)? I have tried various combo of `push_root`, `note_finished_extrinsic` and apply extrinsic, no luck yet.

Yes, this is it. Also make sure that when you build the proof (from BN state), you also run this combo.

**Comment by liuchengxu** [comment]
Somehow the way of adding a new API returning the storage root instead of `ApplyExtrinsicResult` produces a really confusing problem, if I comment out some log containing the `storage_root` call, it does not return the post_execution_root but still return the root before the execution. To reproduce:

```bash
$ cd /tmp
$ git clone https://github.com/subspace/subspace --branch fraud-proof-impl
# Test passes
$ cd cumulus/client/cirrus-executor && RUST_LOG=runtime::executive=debug cargo test fraud_proof -- --nocapture
```

If I comment out https://github.com/subspace/subspace/blob/18431b542e/cumulus/pallets/executive/src/lib.rs#L363-L367 and rerun `RUST_LOG=runtime::executive=debug cargo test fraud_proof -- --nocapture`, the test failed.

```diff
diff --git a/cumulus/pallets/executive/src/lib.rs b/cumulus/pallets/executive/src/lib.rs
index c207758a..d84624b3 100644
--- a/cumulus/pallets/executive/src/lib.rs
+++ b/cumulus/pallets/executive/src/lib.rs
@@ -360,11 +360,11 @@ impl<
                        COnRuntimeUpgrade,
                >::apply_extrinsic(uxt);
 
-               frame_support::log::info!(
-                       target: "cirrus::runtime::executive",
-                       "[apply_extrinsic] after: {:?}",
-                       Block::Hash::decode(&mut Self::storage_root().as_slice()).unwrap()
-               );
+               // frame_support::log::info!(
+                       // target: "cirrus::runtime::executive",
+                       // "[apply_extrinsic] after: {:?}",
+                       // Block::Hash::decode(&mut Self::storage_root().as_slice()).unwrap()
+               // );
 
                res
        }
```

I really don't know why. Could somewhat be related to the underlying `sp_io::storage::root()`? @cheme

**Comment by cheme** [comment]
It does not make sense indeed, I did not check what happen in code (would need substrate branch to check), but my only hypothesis that could make sense is that your call to storage_root in apply extrinsic is called before instruction such as `note_extrinsic` (so some reads are done after the call) and calling it a second time in the log helps completing the proof.

I realize, the different points here are a bit scattered, so I am summing up the time line as I see it:

* transaction pool:
 nothing change, extrinsic is run, extrinsic do not include calls to storage root (can but it will be extrinsic inner logic).

* block import or standard execution (into `execute_block` on state B):

1.  we are at state BN (with some overlay content).
2.  run storage root and store the root of BN: this is needed to get root in state
3.  run `apply_extrinsic_with_len` (there is a few additional changes like `note_extrinsic`).
4.  we are at state BN+1

* Recording proof.

1. on state BN with empty overlay
2. run storage root and store the root of BN: this is needed to get root in state
3. run `apply_extrinsic_with_len` (there is a few additional changes other than the extrinsic like `note_extrinsic`).
4. we are at state BN+1
5. run storage root: this is needed to include content to proof for verification

* Verifying proof

same as registering proof with latest storage root call result being return


An alternative could be to pass storage root when registering and verifying instead of calling first storage_proof, but
storage_proof on empty overlay is close to free.

**Comment by liuchengxu** [comment]
>There is definitely a time when transaction is not a opaque associated type.
When running import it is in client/db code when writing the delta, in test I am not sure. In your case it is possible you still need to call into_storage_proof a bit higher on the calling stack. It is also possible that with the code you use the transaction/overlay payload is dropped before and some changes are required.
Also the concrete struct may not be a PrefixedMemoryDb but a MemoryDb, using memory db in the overlay will work (just ignore prefix parameter for get). May be another type, but should be doable to rebuild the db from it.

I really don't see a way to call into_storage_proof a bit higher on the calling stack in my case. But adding a more trait bound to `B::Transaction` will be helpful and we can get it for free as MemoryDB already implements it. With this addition trait bound and [changing the delta to be a generic parameter](https://github.com/subspace/subspace/commit/ad7649bbcc9c2d103f98437eb72b92a6896fa20d), we no longer have the trait type and concrete struct issue.

```diff
diff --git a/primitives/state-machine/src/backend.rs b/primitives/state-machine/src/backend.rs
index 8d0ac2d136..a08bc53dda 100644
--- a/primitives/state-machine/src/backend.rs
+++ b/primitives/state-machine/src/backend.rs
@@ -37,7 +37,7 @@ pub trait Backend<H: Hasher>: sp_std::fmt::Debug {
        type Error: super::Error;

        /// Storage changes to be applied if committing
-       type Transaction: Consolidate + Default + Send;
+       type Transaction: Consolidate + Default + Send + sp_trie::HashDBT<H, sp_trie::DBValue>;

        /// Type of trie backend storage.
        type TrieBackendStorage: TrieBackendStorage<H, Overlay = Self::Transaction>;
```

@cheme Is it okay for you to simply add this trait bound? If yes, along with the `overlay`, I think they are all required substrate changes to support the extrinsic proof since [the extrinsic execution proof won't be necessarily included in Substate for now](https://github.com/paritytech/substrate/pull/11019#issuecomment-1068465810).

**Comment by cheme** [comment]
It breaks a bit the design.
Would adding something like
```
sc_client_api::backend::StateBackendFor<Backend, Block>::Transaction : sp_trie::HashDBT<H, sp_trie::DBValue>,
```
here https://github.com/subspace/subspace/blob/ad7649bbcc9c2d103f98437eb72b92a6896fa20d/cumulus/client/cirrus-executor/src/lib.rs#L107 works.
Or is there other location this is needed?

**Comment by liuchengxu** [comment]
@cheme Thanks for the clue! That works! Now there is only one mysterious problem in https://github.com/paritytech/substrate/pull/10922#issuecomment-1069100760, I'll try to comprehend your comment and dig into it :D

File: primitives/api/proc-macro/src/impl_runtime_apis.rs
Comment (bkchr) [inline primitives/api/proc-macro/src/impl_runtime_apis.rs:311]:
```suggestion
			fn overlay(&self) -> #crate_::OverlayedChanges {
				::std::clone::Clone::clone(&::std::cell::RefCell::borrow(&self.changes))
```

**Review by bkchr — APPROVED** [review]

**Comment by bkchr** [comment]
@liuchengxu could you please merge master to make the CI happy?

**Comment by cheme** [comment]
@liuchengxu , I lost track of where exposing the overlay is needed, no problem with it, but I just can't remember.

**Comment by liuchengxu** [comment]
@cheme I need the overlay to rebuild the state to re-execute a certain extrinsic precisely, which is basically parent_state + overlay. Specifically, the overlay will be converted to a memory DB(something that implements HashDB) and used to build a `DeltaBackend` in #11019.

**Comment by cheme** [comment]
> @cheme I need the overlay to rebuild the state to re-execute a certain extrinsic precisely, which is basically parent_state + overlay. Specifically, the overlay will be converted to a memory DB(something that implements HashDB) and used to build a `DeltaBackend` in #11019.

Isn't it build from 'transaction' and 'root' field of the result of 'into_storage_changes'?
Or in case of call to `prove_on_trie_backend` it is a mutable reference as input parameter, so we already got it.
(I think I am missing something)

**Comment by liuchengxu** [comment]
@cheme Yes, `transaction` and `transaction_storage_root` is built by `overlay.into_storage_changes()`, but the overlay is built by partially executing a custom block builder, so we need to access the overlay some time during the block execution, this piece of code might be helpful https://github.com/subspace/subspace/blob/0adb26f311/cumulus/client/block-builder/src/lib.rs#L233

**Comment by cheme** [comment]
Why do you need to prepare an overlay with content? I mean everything should be executed on empty overlay (since we now use real states BN being build over trie storage that contains the changes from the overlay)

File: primitives/api/proc-macro/src/impl_runtime_apis.rs
Comment (bkchr) [inline primitives/api/proc-macro/src/impl_runtime_apis.rs:306]:
```suggestion
				::std::clone::Clone::clone(&::std::cell::RefCell::borrow(&self.changes))
```

I have added them on purpose ;) I know that they are not required anymore, but I wanted to ensure that we don't use any local scope.

Comment (liuchengxu) [reply primitives/api/proc-macro/src/impl_runtime_apis.rs]: Oops, I thought that was an accident 🙈

**Comment by liuchengxu** [comment]
Proving the execution `prove_execution_on_trie_backend` will be run on an empty overlay, in order to do that, we need a true state of BN which is achieved by using a `DeltaBackend` that wraps parent_state and overlay.

**Comment by cheme** [comment]
> Proving the execution `prove_execution_on_trie_backend` will be run on an empty overlay, in order to do that, we need a true state of BN which is achieved by using a `DeltaBackend` that wraps parent_state and overlay.

Isn't `DeltaBackend` built from the resul to `into_storage_root` (transaction and root)? (like https://github.com/subspace/subspace/blob/ad7649bbcc9c2d103f98437eb72b92a6896fa20d/cumulus/client/cirrus-executor/src/lib.rs#L330)

**Comment by liuchengxu** [comment]
Yes, it's built from the result of `into_storage_changes`. The test https://github.com/subspace/subspace/blob/ad7649bbcc/cumulus/client/cirrus-executor/src/tests.rs#L169-L179, creating the DetaBackend https://github.com/subspace/subspace/blob/ad7649bbcc/cumulus/client/fraud-proof/src/lib.rs#L79

**Comment by cheme** [comment]
So I don't understand where we need to access the overlay with the new method :(

**Comment by liuchengxu** [comment]
https://github.com/subspace/subspace/blob/ad7649bbcc/cumulus/client/block-builder/src/lib.rs#L233 This line requires this new `overlay` API, otherwise, we can't get the overlayed changes before executing the extrinsic N, thus we can't build the state BN later. :P

**Comment by cheme** [comment]
Brings me back to https://github.com/paritytech/substrate/pull/10922#issuecomment-1072578687 

> we can't get the overlayed changes before executing the extrinsic N, thus we can't build the state BN later. :P

but state bn is build from executing from state B all on init, and all extrinsc to N. So run on empty overlay. And then we call into_storage_change.

**Comment by liuchengxu** [comment]
>but state bn is build from executing from state B the block up to extrinsic N. So it run on empty overlay. And then we extract BN by using into_storage_change.

Oh, get your point finally! Yes, this PR is unnecessary! I was solving the problem indirectly. [This](https://github.com/subspace/subspace/commit/186c1a8a2f83298b7befb8b988cf2ce6b6a9ad67) can be useful to understand what happens here for these interested in this issue.

@cheme Really appreciated about all the amazing help on this topic! Hope I can figure out the last issue in https://github.com/paritytech/substrate/pull/10922#issuecomment-1068997467 soon :P

**Comment by cheme** [comment]
No problem, I think I still got email alert on closed PR, don't hesitate to continue the conversation here if needed, it is an interesting topic.

**Pull request closed without merging**
